### PR TITLE
Simplification pass for ACIR

### LINKED ISSUE #573: Optimise away predicates which are always true
# Problem

The following program:

```rust
 fn main( y: Field, z: Field) {
    let x = y + z;
    constrain x == 20;
}
```

Will generate the following opcodes:

```
1x1 + 1x2 + -1x3 + -20 = 0
x4=1/x3, or 0
1x3*x4 + -1x5 + 0 = 0
1x3*x5 + -1x3 + 0x5 + 0 = 0
1x5 + 0 = 0
```

Note that there is an inversion being done here, when the original code should not require it. This is because the predicate is not being optimised away

# Solution

Add optimisation to remove this extra overhead.

# Alternatives considered

(Describe any alternative solutions you have considered.)

# Additional context

### PR DESCRIPTION
**Pull request by guipublic**
# Related issue(s)

Resolves #573 

# Description

## Summary of changes

Add a simplification pass which removes gates of the form `x=field value`, by replacing x by its value in all the other gates.
The simplification is done on the fly, every-time a gate is added to the circuit. If it can be simplified, we update the previous gates using the solved value until nothing is solved. Usually only one pass is needed (so two in total to see that the second one did not solve anything)

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

### CODE REVIEW

**Comment by kevaundray** [comment]
For reference, why is this done on the fly instead of at the end like the other passes? 

(Doing it on the fly seems to have made a lot of the code now need to take a Result)

File: crates/acvm/src/simplify.rs [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    Replace(Gate),
+    SolvedGate,
+    SolvedWitness(Witness),
+    UnsatisfiedConstrain(usize),
+}
+
+pub struct CircuitSimplifier {
+    abi_len: u32,                                //number of witness in the ABI
+    pub solved: BTreeMap<Witness, FieldElement>, //pub is TEMP
+    pub defined: HashSet<Witness>, //list des solved witness ayant une gate pour les definir
```
Comment (kevaundray) [inline crates/acvm/src/simplify.rs:29]: this is french btw

Comment (guipublic) [reply crates/acvm/src/simplify.rs]: Done

**Comment by kevaundray** [comment]
There's been quite a bit of merge conflicts on this, that I will resolve

**Comment by jfecher** [comment]
What is the status on this PR? Do we want to try to fix the conflicts and merge it, restart it, or abandon it?

**Comment by guipublic** [comment]
> What is the status on this PR? Do we want to try to fix the conflicts and merge it, restart it, or abandon it?

I think the acvm changes are too big for solving the conflicts. I will create another PR (probably we need two now) and then we can abandon this one.

**Comment by TomAFrench** [comment]
Closing in favour of https://github.com/noir-lang/acvm/pull/151

**Pull request closed without merging**